### DISCUSSION TITLE
Does xmake-repo support package dependency conflict resolution?

### DISCUSSION THREAD

**Discussion by calebkiage** [xmake-io/xmake#3887] [Q&A] [answered]
Does xmake support dependency version conflict resolution [like conan](https://docs.conan.io/2/tutorial/versioning/conflicts.html)?

I'm facing a strange issue where the following requires add 2 different fmt libraries to the lock file.  Can I disable that?
```lua
add_requires("fmt 9", "spdlog 1.11", { configs = { fmt_external = true }})
```

Lock file:
```
{
    __meta__ = {
        version = "1.0"
    },
    ["windows|x64"] = {
        ["fmt 9#8274c4fc"] = {
            repo = {
                branch = "master",
                commit = "5076dd50345ce1c9347f6f51ee546a18db035e98",
                url = "https://gitlab.com/tboox/xmake-repo.git"
            },
            version = "9.1.0"
        },
        ["fmt#31fecfc4"] = {
            repo = {
                branch = "master",
                commit = "5076dd50345ce1c9347f6f51ee546a18db035e98",
                url = "https://gitlab.com/tboox/xmake-repo.git"
            },
            version = "10.0.0"
        },
        ["spdlog 1.11#8274c4fc"] = {
            repo = {
                url = "dependencies"
            },
            version = "v1.11.0"
        }
    }
}
```

**Comment by waruqi** [comment, marked as answer]
xmake has a dependency conflict check, if there are no errors reported, there is no conflict. This is because different versions of the same package are allowed to be used for different targets.

Only if there are multiple versions of the same package on a dependency link will there be a conflict. `A -> B 1.1 -> C -> B 1.0`

> add_requires("fmt 9", "spdlog 1.11", { configs = { fmt_external = true }})

You are currently using a package configuration that does not have any conflicts.

```
- fmt 9
- spdlog -> fmt  10
```

If you want to keep their versions consistent, you can solve that with add_requireconfs as well.

```
add_requireconfs("spdlog.fmt", {version = "9", override = true})
```

**Reply by calebkiage** [reply]
This doesn't seem to work the way I expect.

The lock file still has 2 fmt libraries:

```
...
        ["fmt 9#8274c4fc"] = {
            repo = {
                branch = "master",
                commit = "5076dd50345ce1c9347f6f51ee546a18db035e98",
                url = "https://gitlab.com/tboox/xmake-repo.git"
            },
            version = "9.1.0"
        },
        ["fmt#31fecfc4"] = {
            repo = {
                branch = "master",
                commit = "5076dd50345ce1c9347f6f51ee546a18db035e98",
                url = "https://gitlab.com/tboox/xmake-repo.git"
            },
            version = "9.1.0"
        }
...
```

This wouldn't be a problem except that the linker adds both libraries:

```
link.exe -nologo ...  fmtd.lib fmt.lib -out:build\windows\x64\debug\test.exe ...
```

**Reply by calebkiage** [reply]
I think I found the issue.
I had specified a requireconfs for all packages specifying a debug config:

```
add_requireconfs("*", { configs = { debug =true }})
```

Adding the debug option to spdlog's conf fixed the dependencies

```
add_requireconfs("spdlog.fmt", {version = "9", override = true, configs = { debug = true  })
```

I also had to separate the fmt and spdlog add_requires:

```
add_requires("fmt 9")
add_requires("spdlog 1.11", { configs = { fmt_external = true }})
```
